src/net.h

```diff
@@ -114,6 +114,15 @@ struct CSerializedNetMsg
     std::string m_type;
 };
 
+const std::vector<std::string> CONNECTION_TYPE_DOC{
```

Comment:
Is it possible to add an assert that the size of this vector is the same as the size of ConnectionType, so it's not possible to update one without the other?

Reply:
I like the idea, but it looks like there's no straightforward way to get the length of the enum? :(

Reply:
I believe you can't in general.

But you can add a "MAX" element to the enum, which is the length.

If you make the list an std::array instead of std::vector, you can make it a static_assert with that even.

Reply:
I'm unclear on what's being recommended here. 

Should I update to make `CONNECTION_TYPE_DOC` an array, allocate size based on `MAX` from enum, then hardcode the `MAX` value into the static assert to compare it to the size of array?

Reply:
Just in theory it would be possible to get the number of enum values via the preprocessor:


```diff
diff --git a/src/util/enum_class.h b/src/util/enum_class.h
new file mode 100644
index 0000000000..90aefe8c3b
--- /dev/null
+++ b/src/util/enum_class.h
@@ -0,0 +1,14 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_ENUM_CLASS_H
+#define BITCOIN_UTIL_ENUM_CLASS_H
+
+#include <boost/preprocessor/variadic/size.hpp>
+
+#define ENUM_CLASS(name, ...)        \
+    enum class name { __VA_ARGS__ }; \
+    constexpr size_t name##_##SIZE { BOOST_PP_VARIADIC_SIZE(__VA_ARGS__) }
+
+#endif // BITCOIN_UTIL_ENUM_CLASS_H
```

This could be used to write

```cpp
static_assert(CONNECTION_TYPE_DOC.size() == ConnectionType_SIZE, "inconsistent docs");
```

Probably not worth the effort. Just saying that it is possible.

Reply:
bah. Adding a `MAX` value breaks all the (important) "value not handled in switch statement" warnings. I don't think this is worth it. Sorry for the noise!

Reply:
@jnewbery this works, but also meh
```cpp
enum class E {
  X,
  Y
  LAST = Y
};

int f(E e) {
  switch (e) {
  case E::X: return 1;
  case E::Y: return 2;
  }
}
```

Reply:
> also meh

I agree!